modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBTrainer.java

```diff
@@ -0,0 +1,190 @@
+            @Override public Double apply(Vector features) {
+                return internalLabelToExternal(super.apply(features));
+            }
+        };
+    }
+
+    /**
+     * Defines unique labels in dataset if need (useful in case of classification).
+     *
+     * @param builder Dataset builder.
+     * @param featureExtractor Feature extractor.
+     * @param lExtractor Labels extractor.
+     */
+    protected abstract  <V, K> void learnLabels(DatasetBuilder<K, V> builder,
+        IgniteBiFunction<K, V, double[]> featureExtractor, IgniteBiFunction<K, V, Double> lExtractor);
+
+    /**
+     * Returns regressor model trainer for one step of GDB.
+     */
+    @NotNull protected abstract DatasetTrainer<? extends Model<Vector, Double>, Double> buildBaseModelTrainer();
+
+    /**
+     * Maps external representation of label to internal.
+     *
+     * @param lbl Label value.
+     */
+    protected abstract double externalLabelToInternal(double lbl);
+
+    /**
+     * Maps internal representation of label to external.
+     *
+     * @param lbl Label value.
+     */
+    protected abstract double internalLabelToExternal(double lbl);
+
+    /**
+     * Compute mean value of label as first approximation.
+     *
+     * @param builder Dataset builder.
+     * @param featureExtractor Feature extractor.
+     * @param lbExtractor Label extractor.
+     */
+    protected <V, K> IgniteBiTuple<Double, Long> computeInitialValue(DatasetBuilder<K, V> builder,
+        IgniteBiFunction<K, V, double[]> featureExtractor,
+        IgniteBiFunction<K, V, Double> lbExtractor) {
+
+        try (Dataset<EmptyContext, DecisionTreeData> dataset = builder.build(
+            new EmptyContextBuilder<>(),
+            new DecisionTreeDataBuilder<>(featureExtractor, lbExtractor)
+        )) {
+            IgniteBiTuple<Double, Long> meanTuple = dataset.compute(
+                data -> {
+                    double sum = Arrays.stream(data.getLabels()).map(this::externalLabelToInternal).sum();
+                    return new IgniteBiTuple<>(sum, (long)data.getLabels().length);
+                },
+                (a, b) -> {
+                    if (a == null)
+                        return b;
+                    if (b == null)
+                        return a;
+
+                    a.set1(a.get1() + b.get1());
+                    a.set2(a.get2() + b.get2());
+                    return a;
+                }
+            );
+
+            meanTuple.set1(meanTuple.get1() / meanTuple.get2());
+            return meanTuple;
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Estimate gradient value for answer-prediction pair.
+     *
+     * @param sampleSize Sample size.
+     * @param answer Right answer.
+     * @param prediction Prediction.
+     */
+    protected abstract double grad(long sampleSize, double answer, double prediction);
```

Comment:
I suggest use here composition instead of inheritance. Let's feed grad as lambda as one more constructor parameter. This way we'll be able to make for example more easy grid search on losses.

Reply:
ok.